R/progress.R

```diff
@@ -0,0 +1,78 @@
+progress_bar <- function(n_chains, n_steps, progress, call = NULL) {
+  progress <- show_progress_bar(progress, call)
+  if (progress) {
+    progress_bar_detail(n_chains, n_steps)
+  } else {
+    progress_bar_null()
+  }
+}
+
+
+show_progress_bar <- function(progress, call = NULL) {
+  if (is.null(progress)) {
+    progress <- getOption("mcstate2.progress", TRUE)
+  }
+  ## Error here is not great if we get this from an option; but need
+  ## to disable quoting there in some cases.
+  assert_scalar_logical(progress, call = call)
+}
+
+
+## This will be the case to use where we can report back in detail
+## about what is running (for each chain we can report about the
+## progress within that chain).  The other likely option will be a
+## very coarse detail which is the state of each chain only; that
+## might be the best we can do for some of the other parallel
+## backends, but we'll see how packages that implement progress bars
+## there cope.
+progress_bar_detail <- function(n_chains, n_steps) {
+  e <- new.env()
+  e$n <- rep(0, n_chains)
+  overall <- progress_overall(n_chains, n_steps)
+  fmt <- paste("Sampling {overall(e$n)} {cli::pb_bar} |",
+               "{cli::pb_percent} ETA: {cli::pb_eta}")
+  id <- cli::cli_progress_bar(
+    total = n_chains * n_steps,
+    format = fmt,
+    .auto_close = FALSE)
+
+  function(chain_index) {
+    function(at) {
+      e$n[[chain_index]] <- at
+      cli::cli_progress_update(id = id, set = sum(e$n))
+    }
+  }
+}
+
+
+## Dummy version that can be used where no progress bar is wanted.
+progress_bar_null <- function(...) {
+  function(chain_index) {
+    function(at) {
+    }
+  }
+}
+
+
+progress_overall <- function(n_chains, n_steps) {
+  if (n_chains == 1) {
+    return(function(n) "")
+  }
+  sym <- unlist(cli::symbol[paste0("lower_block_", 1:8)], use.names = FALSE)
```

Comment:
In R from windows command prompt, cli::symbol$lower_block 4 and 8 work; the others do not.
Normal underscore is close to $lower_block_1 but one pixel higher.

You could consider doubling up on ascii codes 176, 177, 178, 219 as a windows-specific best try
░░▒▒▓▓██

In RStudio on Win it is a bit better, but some horizontal spacing variation makes it jiggly

Reply:
Another option would be to use the fallback characters? I know cli degrades to something like `_.oO#` which could work.  It might be worth checking to see if this is a font issue too, and/or logging it as an issue in cli as I think the intent is that this should be something that package would smooth over and not us